protocols/mdns/src/behaviour.rs

```diff
@@ -36,29 +36,53 @@ use smallvec::SmallVec;
 use std::collections::hash_map::{Entry, HashMap};
 use std::{cmp, fmt, io, net::IpAddr, pin::Pin, task::Context, task::Poll, time::Instant};
 
-#[cfg(feature = "async-io")]
-use crate::behaviour::{socket::asio::AsyncUdpSocket, timer::asio::AsyncTimer};
+/// A Fallible Default trait. Used on [`GenMdns`] to Allow accepting `IfWatcher`'s.
+pub trait TryDefault: Sized {
+    /// Fallible equivalent of [`Default::default`].
+    fn try_default() -> Result<Self, std::io::Error>;
+}
 
 /// The type of a [`GenMdns`] using the `async-io` implementation.
 #[cfg(feature = "async-io")]
-pub type Mdns = GenMdns<AsyncUdpSocket, AsyncTimer>;
+pub mod async_io {
+    use super::{GenMdns, TryDefault};
+    use crate::behaviour::{socket::asio::AsyncUdpSocket, timer::asio::AsyncTimer};
+    use if_watch::smol::IfWatcher;
 
-#[cfg(feature = "tokio")]
-use crate::behaviour::{socket::tokio::TokioUdpSocket, timer::tokio::TokioTimer};
+    pub type Mdns = GenMdns<AsyncUdpSocket, AsyncTimer, IfWatcher>;
+
+    impl TryDefault for IfWatcher {
+        fn try_default() -> Result<Self, std::io::Error> {
+            Self::new()
+        }
+    }
+}
 
 /// The type of a [`GenMdns`] using the `tokio` implementation.
 #[cfg(feature = "tokio")]
-pub type TokioMdns = GenMdns<TokioUdpSocket, TokioTimer>;
+pub mod tokio {
+    use super::{GenMdns, TryDefault};
+    use crate::behaviour::{socket::tokio::TokioUdpSocket, timer::tokio::TokioTimer};
+    use if_watch::tokio::IfWatcher;
+
+    pub type Mdns = GenMdns<TokioUdpSocket, TokioTimer, IfWatcher>;
+
+    impl TryDefault for IfWatcher {
+        fn try_default() -> Result<Self, std::io::Error> {
+            Self::new()
+        }
+    }
+}
 
 /// A `NetworkBehaviour` for mDNS. Automatically discovers peers on the local network and adds
 /// them to the topology.
 #[derive(Debug)]
-pub struct GenMdns<S, T> {
+pub struct GenMdns<S, T, W> {
     /// InterfaceState config.
     config: MdnsConfig,
 
     /// Iface watcher.
-    if_watch: IfWatcher,
+    if_watch: W,
```

Comment:
Can we do `T::IfWatcher` instead, similar to what @rkuhn did for TCP?

Reply:
Yeah, great suggestion! Ptal if that's what you were expecting Thomas.